pkg/handler/storage/client.go

```diff
@@ -0,0 +1,427 @@
+	}
+
+	if err := rbac.AllowProjectScope(ctx, "region:filestorage:v2", identityapi.Read, result.Labels[coreconstants.OrganizationLabel], result.Labels[coreconstants.ProjectLabel]); err != nil {
+		return nil, err
+	}
+
+	// Only allow access to resources created by this API (temporarily).
+	v, ok := result.Labels[constants.ResourceAPIVersionLabel]
+	if !ok {
+		return nil, errors.HTTPNotFound()
+	}
+
+	version, err := constants.UnmarshalAPIVersion(v)
+	if err != nil {
+		return nil, errors.OAuth2ServerError("unable to parse API version")
+	}
+
+	if version != 2 {
+		return nil, errors.HTTPNotFound()
+	}
+
+	return result, nil
+}
+
+func (c *Client) Get(ctx context.Context, storageID string) (*openapi.StorageV2Read, error) {
+	result, err := c.GetRaw(ctx, storageID)
+	if err != nil {
+		return nil, err
+	}
+
+	return convertV2(result), nil
+}
+
+func (c *Client) generateV2(ctx context.Context, organizationID, projectID, regionID string, request *openapi.StorageV2Update) (*regionv1.FileStorage, error) {
+
+	size, err := convertSize(request.Spec.Size)
+	if err != nil {
+		return nil, errors.HTTPNotFound()
+	}
+	out := &regionv1.FileStorage{
+		ObjectMeta: conversion.NewObjectMetadata(&request.Metadata, c.namespace).
+			WithOrganization(organizationID).
+			WithProject(projectID).
+			WithLabel(constants.RegionLabel, regionID).
+			WithLabel(constants.ResourceAPIVersionLabel, constants.MarshalAPIVersion(2)).
+			Get(),
+		Spec: regionv1.FileStorageSpec{
+			Tags:        conversion.GenerateTagList(request.Metadata.Tags),
+			Size:        size,
+			Attachments: convertAttachmentList(request.Spec.Attachments),
+		},
+	}
+
+	if err := util.InjectUserPrincipal(ctx, organizationID, projectID); err != nil {
+		return nil, errors.OAuth2ServerError("unable to set principal information").WithError(err)
+	}
+
+	if err := common.SetIdentityMetadata(ctx, &out.ObjectMeta); err != nil {
+		return nil, errors.OAuth2ServerError("failed to set identity metadata").WithError(err)
+	}
+
+	return out, nil
+}
+
+func convertAttachmentList(in *openapi.StorageAttachmentV2Spec) []regionv1.Attachment {
+	out := make([]regionv1.Attachment, len(in.NetworkIds))
+	for i := range in.NetworkIds {
+		out[i] = regionv1.Attachment{
+			NetworkID: in.NetworkIds[i].Id,
+		}
+	}
+	return out
+}
+
+func convertSize(size string) (resource.Quantity, error) {
+	return resource.ParseQuantity(size)
+}
+
+type createSaga struct {
+	client  *Client
+	request *openapi.StorageV2Create
+
+	filestorage *regionv1.FileStorage
+}
+
+func newCreateSaga(client *Client, request *openapi.StorageV2Create) *createSaga {
+	return &createSaga{
+		client:  client,
+		request: request,
+	}
+}
+
+// createAllocation creates an allocation for the storage ID and then attaches
+// the allocation ID to the network for persistence.
+func (s *createSaga) createAllocation(ctx context.Context) error {
+	required := identityapi.ResourceAllocationList{
+		{
+			Kind:      "filestorage",
```

Comment:
We'll need to make sure there's quota metadata for this. I don't actually know where those come from 😳